### PR TITLE
Add fully Bayesian modelling via NUTS

### PR DESCRIPTION
**Pull request by jacobrgardner**
Alright, the time came to take matters in to our own hands on this :-)

This is mostly done in my mind, two things left to do that I'll finish hopefully today or tomorrow:
- [x] Add `expand` methods to the rest of the priors (100% necessary because of how extending torch Distributions work).
- [x] Add a few more custom `pyro_load_from_samples` override to do additional cleanup. These will all be two liners that do one thing and call super. For example, we need to update `batch_shape` in the kernels.
- [x] Unit tests + docs + text explaining stuff in the notebook.

The interface is pretty clean in my opinion (see the example notebook). Here's sampling:
```python
model = ExactGPModel(train_x, train_y, likelihood)

# Register normal GPyTorch priors
model.mean_module.register_prior("mean_prior", UniformPrior(-1, 1), "constant")
model.covar_module.base_kernel.register_prior("lengthscale_prior", UniformPrior(0.01, 0.2), "lengthscale")
model.covar_module.register_prior("outputscale_prior", UniformPrior(1, 2), "outputscale")
likelihood.register_prior("noise_prior", LogNormalPrior(-1.5, 0.1), "noise")

mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

def pyro_model(x, y):
    model.pyro_sample_from_prior()
    output = model(x)
    loss = mll.pyro_factor(output, y)
    return y

nuts_kernel = NUTS(pyro_model, adapt_step_size=True)
mcmc_run = MCMC(nuts_kernel, num_samples=100, warmup_steps=200)
mcmc_run.run(train_x, train_y)
```

And here's evaluation:
```python
model.pyro_load_from_samples(mcmc_run.get_samples())
model.eval()
expanded_train_x = train_x.unsqueeze(-1).repeat(100, 1, 1)
print(expanded_train_x.shape)
with gpytorch.settings.debug(False):
    output = model(expanded_train_x)
```

@Balandat @gpleiss @dme65

### CODE REVIEW

**Comment by Balandat** [comment]
cc @qingfeng10

**Comment by jacobrgardner** [comment]
Test failures are just the incomplete Distribution/Prior refactor.

File: gpytorch/kernels/rbf_kernel.py [hunk 1]
```diff
@@ -73,7 +73,7 @@ def forward(self, x1, x2, diag=False, **params):
         if (
             x1.requires_grad
             or x2.requires_grad
-            or (self.ard_num_dims is not None and self.ard_num_dims > 1)
+            or self.lengthscale.numel() > 1
```
Comment (Balandat) [inline gpytorch/kernels/rbf_kernel.py]: This is not equivalent if `batch_shape` is not set. Is this intended?

File: gpytorch/mlls/marginal_log_likelihood.py [hunk 2]
```diff
@@ -31,3 +31,9 @@ def forward(self, output, target):
         - target: (Variable) - the target values
         """
         raise NotImplementedError
+
+    def pyro_factor(self, output, target):
+        """
+        As forward, but register the loss with pyro using the pyro.factor primitive.
+        """
+        raise NotImplementedError
```
Comment (Balandat) [inline gpytorch/mlls/marginal_log_likelihood.py]: newline

File: gpytorch/models/exact_gp.py [hunk 3]
```diff
@@ -45,6 +45,15 @@ def _apply(self, fn):
             self.train_targets = fn(self.train_targets)
         return super(ExactGP, self)._apply(fn)
 
+    def pyro_load_from_samples(self, samples_dict):
+        # Pyro always puts the samples in the first batch dimension
+        num_samples = next(iter(samples_dict.values())).size(0)
+        self.train_inputs = tuple(
+            tri.unsqueeze(0).repeat(num_samples, *([1] * len(tri.shape))) for tri in self.train_inputs
```
Comment (Balandat) [inline gpytorch/models/exact_gp.py]: do you want to repeat here or would expand do as well?

`tri.expand(num_samples, *tri.shape)`

File: gpytorch/models/exact_gp.py [hunk 4]
```diff
@@ -45,6 +45,15 @@ def _apply(self, fn):
...
+        )
+        self.train_targets = self.train_targets.unsqueeze(0).repeat(num_samples, *([1] * len(self.train_targets.shape)))
```
Comment (Balandat) [inline gpytorch/models/exact_gp.py]: same as above

File: gpytorch/priors/torch_priors.py [hunk 5]
```diff
@@ -23,6 +24,34 @@ def __init__(self, loc, scale, validate_args=False, transform=None):
         _bufferize_attributes(self, ("loc", "scale"))
         self._transform = transform
 
+    def expand(self, batch_shape):
+        return Normal.expand(self, batch_shape, _instance=self)
```
Comment (Balandat) [inline gpytorch/priors/torch_priors.py]: kind of annoying  how you can't call the superclass method here ...

File: gpytorch/kernels/rbf_kernel.py (hunk 1 above)
Comment (jacobrgardner) [reply gpytorch/kernels/rbf_kernel.py]: This will probably get reverted in the clean up -- the Kernel will have to override `pyro_load_from_samples` anyways to update batch_shape. Right now, if batch_shape is wrong, you can only forward through the model with debug mode off.

File: gpytorch/models/exact_gp.py (hunk 3 above)
Comment (jacobrgardner) [reply gpytorch/models/exact_gp.py]: Yeah, expand should work

File: gpytorch/priors/torch_priors.py (hunk 5 above)
Comment (jacobrgardner) [reply gpytorch/priors/torch_priors.py]: Agreed. Unfortunately, `torch.distributions.Distribution` explicitly forbids it. I've no idea why, but it means we need to define `expand` methods for all of our priors that are just a 1 line boilerplate.

The other thing we'll need to add to priors is `rsample` functionality for any priors we want to use for sampling -- e.g., I define a `UniformPrior` in this PR mostly because `SmoothedBoxPrior` doesn't have an rsample yet.

File: gpytorch/module.py [hunk 6]
```diff
@@ -270,6 +278,15 @@ def sample_from_prior(self, prior_name):
             raise RuntimeError("Must provide inverse transform to be able to sample from prior.")
         setting_closure(prior.sample())
 
+    def pyro_sample_from_prior(self):
+        return _pyro_sample_from_prior(module=self, memo=None, prefix="")
+
+    def pyro_load_from_samples(self, samples_dict):
+        self._strict(False)
```
Comment (jacobrgardner) [inline gpytorch/module.py]: This needs to call recursively rather than do everything from the top level, because some modules will have additional behavior (e.g., kernels will have to update batch_shape).

File: gpytorch/priors/torch_priors.py (hunk 5 above)
Comment (Balandat) [reply gpytorch/priors/torch_priors.py]: > Unfortunately, torch.distributions.Distribution explicitly forbids it. I've no idea why, but it means we need to define expand methods for all of our priors that are just a 1 line boilerplate.

Yeah I noticed that as well. I guess you could make a `LessOpinionatedTorchDistributionMixin` that reimplements [this](https://github.com/pytorch/pytorch/blob/master/torch/distributions/distribution.py#L255-L260) and removes the error there :)

**Comment by jacobrgardner** [comment]
@Balandat @gpleiss @dme65 This is ready for review.

File: gpytorch/mlls/exact_marginal_log_likelihood.py [hunk 7]
```diff
@@ -37,3 +37,9 @@ def forward(self, output, target, *params):
         # Scale by the amount of data we have
         num_data = target.size(-1)
         return res.div_(num_data)
+
+    def pyro_factor(self, output, target, *params):
+        import pyro
+        loss = self(output, target, *params)
+        pyro.factor("gp_mll", loss)
```
Comment (martinjankowiak) [inline gpytorch/mlls/exact_marginal_log_likelihood.py]: isn't this the wrong sign?

Comment (jacobrgardner) [reply gpytorch/mlls/exact_marginal_log_likelihood.py]: No, mll returns a positive log likelihood. We negate it manually in all examples (e.g. `loss = -mll(output, train_y)`

File: gpytorch/priors/torch_priors.py [hunk 8]
```diff
@@ -23,6 +24,35 @@ def __init__(self, loc, scale, validate_args=False, transform=None):
...
+        LogNormal.__init__(self, loc=loc, scale=scale, validate_args=validate_args)
+        self._transform = transform
+
+    def expand(self, batch_shape):
+        return LogNormal.expand(self, batch_shape, _instance=self)
+
+
+class UniformPrior(Prior, Uniform):
+    """
+    Log Normal prior.
```
Comment (jacobrgardner) [inline gpytorch/priors/torch_priors.py]:
```suggestion
    Uniform prior.
```

File: gpytorch/mlls/marginal_log_likelihood.py [hunk 9]
```diff
@@ -31,3 +31,9 @@ def forward(self, output, target):
         - target: (Variable) - the target values
         """
         raise NotImplementedError
+
+    def pyro_factor(self, output, target):
+        """
+        As forward, but register the loss with pyro using the pyro.factor primitive.
```
Comment (jacobrgardner) [inline gpytorch/mlls/marginal_log_likelihood.py]:
```suggestion
        As forward, but register the MLL with pyro using the pyro.factor primitive.
```

File: .travis.yml [hunk 10]
```diff
@@ -15,8 +15,10 @@ install:
       pip install numpy;
       pip install --pre torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html;
       python setup.py build develop;
+      pip install pyro-ppl;
     else
       pip install torch==1.3.0+cpu -f https://download.pytorch.org/whl/torch_stable.html;
+      pip install pyro-ppl;
```
Comment (gpleiss) [inline .travis.yml]: Let's remove the changes to the travis.yml in favor of the ones merged in with the variational PR.

File: gpytorch/module.py [hunk 11]
```diff
@@ -270,6 +278,46 @@ def sample_from_prior(self, prior_name):
...
 
+    def pyro_sample_from_prior(self):
+        """
+        For each parameter in this Module and submodule that have defined priors, sample a value for that parameter
+        from its corresponding prior with a pyro.sample primitive and load the resulting value in to the parameter.
+
+        This method can be used in a Pyro model to conveniently define pyro sample sites for all
+        parameters of the model that have GPyTorch priors registered to them.
+        """
+        return _pyro_sample_from_prior(module=self, memo=None, prefix="")
```
Comment (gpleiss) [inline gpytorch/module.py]: Can we add a try/catch ImportError here?
